python/src/kagent/memory/_pinecone_memory.py

```diff
@@ -0,0 +1,178 @@
+import asyncio
+from typing import Any, List, Optional, Sequence
+
+from autogen_core import CancellationToken, Component
+from autogen_core.memory import Memory, MemoryContent, MemoryMimeType, MemoryQueryResult, UpdateContextResult
+from autogen_core.model_context import ChatCompletionContext
+from autogen_core.models import SystemMessage
+from loguru import logger
+from pinecone import Pinecone
+from pinecone.core.openapi.db_data.model.hit import Hit
+from pinecone.data import Index
+from pydantic import BaseModel, Field, SecretStr
+from typing_extensions import Self
+
+
+class PineconeMemoryConfig(BaseModel):
+    api_key: SecretStr = Field(..., description="The API key for the Pinecone API")
+    index_host: str = Field(..., description="The host for the Pinecone index")
+    top_k: int = Field(default=5, description="The number of results to retrieve from Pinecone")
+    namespace: Optional[str] = Field(default=None, description="The Pinecone namespace to query")
+    record_fields: Optional[List[str]] = Field(description="The fields to retrieve from the Pinecone index")
+    score_threshold: float = Field(
+        default=0.0,
+        description="The score threshold of results to include in the context. Results with a score below this threshold will be ignored.",
+    )
+
+
+class PineconeMemory(Memory, Component[PineconeMemoryConfig]):
+    component_config_schema = PineconeMemoryConfig
+    component_type = "memory"
+    component_provider_override = "kagent.memory.PineconeMemory"
+
+    def __init__(self, config: PineconeMemoryConfig):
+        self._config = config
+        self._pc: Pinecone | None = None
+        self._index: Index | None = None
+        self._initialize()
+
+    def _initialize(self):
+        """Initialize Pinecone if not already done."""
+        if self._pc is None:
+            try:
+                self._pc = Pinecone(api_key=self._config.api_key.get_secret_value(), host=self._config.index_host)
+                self._index = self._pc.Index(host=self._config.index_host)
+            except Exception as e:
+                logger.error(f"Failed to initialize Pinecone: {e}")
+                raise Exception(f"Failed to initialize Pinecone: {e}") from e
```

Comment:
Nit: `Pinecone` and `Index` can no longer be `None` since you are initializing or failing